lib/Accessibility.js

```diff
@@ -0,0 +1,413 @@
+/**
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const {helper} = require('./helper');
+
+/**
+ * @typedef {Object} SerializedAXNode
+ * @property {string} role
+ *
+ * @property {string=} name
+ * @property {string|number=} value
+ * @property {string=} description
+ *
+ * @property {string=} keyshortcuts
+ * @property {string=} roledescription
+ * @property {string=} valuetext
+ *
+ * @property {boolean=} disabled
+ * @property {boolean=} expanded
+ * @property {boolean=} focused
+ * @property {boolean=} modal
+ * @property {boolean=} multiline
+ * @property {boolean=} multiselectable
+ * @property {boolean=} readonly
+ * @property {boolean=} required
+ * @property {boolean=} selected
+ *
+ * @property {boolean|"mixed"=} checked
+ * @property {boolean|"mixed"=} pressed
+ *
+ * @property {number=} level
+ * @property {number=} valuemin
+ * @property {number=} valuemax
+ *
+ * @property {string=} autocomplete
+ * @property {string=} hasPopup
+ * @property {string=} invalid
+ * @property {string=} orientation
+ *
+ * @property {Array<SerializedAXNode>=} children
+ */
+
+class Accessibility {
+  /**
+   * @param {!Puppeteer.CDPSession} client
+   */
+  constructor(client) {
+    this._client = client;
+  }
+
+  /**
+   * @return {!Promise<!SerializedAXNode>}
+   */
+  async snapshot({interestingOnly = true} = {}) {
+    const {nodes} = await this._client.send('Accessibility.getFullAXTree');
+    const root = AXNode.createTree(nodes);
+    if (!interestingOnly)
+      return root.serialize(true);
+
+    /** @type {!Map<!AXNode, !SerializedAXNode>} */
+    const axNodeToSerializedNode = new Map();
+    for (const node of root.interestingSubtree()) {
+      const serializedNode = node.serialize(false);
+      axNodeToSerializedNode.set(node, serializedNode);
+      let parent = node.parent();
+      while (parent) {
+        const serializedParent = axNodeToSerializedNode.get(parent);
+        if (serializedParent) {
+          if (!serializedParent.children)
+            serializedParent.children = [];
+          serializedParent.children.push(serializedNode);
+          break;
+        }
+        parent = parent.parent();
+      }
+    }
+    return axNodeToSerializedNode.get(root);
+  }
+}
+
+class AXNode {
```

Comment:
Let's rewrite this to better outline what's going on here.
There are 2 steps:
1. Mark interesting nodes, `const interestingNodes = markInterstingNodes(tree)`. This'll be a tree traversal function
that gets a `parent: AXNode` and an `isInsideControl: boolean` parameters. This seems to be enough to compute interesting nodes.
2. Filter interesting nodes only: `const newTree = filterTree(oldTree, whitelistedNodes)`. This is just a tree filtering primitive.

For this to work, we'd need to split inhertance from AXNodes:
- Have `AXNode` that doesn't have any children/parent
- Have `TreeNode` which has just 2 fields: `axNode: AXNode` and `children: Array<TreeNode>`